### PR TITLE
OSBuildEnvConfig Controller - Create and Setup Workers

### PR DESCRIPTION
**Pull request by ygalblum**
Controller
----------
Create ConfigMaps for Ansible configuration and playbook
Create ConfigMap for the osbuild-worker
Create Secret with SSH Key-Pair if needed
For each VM:
- If needed, create the VM and expose its SSH port as a service
- Create certificate
- Create ConfigMap for the inventory
- Create the Setup Job
Remove the worker certificate secrets on teardown

main.go
-------
Add kubevirt v1 to Scheme
Create Job and VM repositories

Templates
---------
Add templates for:
- Worker VM
- Worker Setup Job
- OSBuild-Worker configuration
- Ansible Config
- Ansible Inventory
- Ansible Playbook

Repository
----------
Add repositories for Job and CNV VM

RBAC
----
Add permissions for Jobs and CNV VMs

SSHKey
------
Add package for SSH Key-Pair generation

Setup Container image
---------------------
Add Dockerfile
Add Makefile rules
Add ansible requirements.yml file

Type
----
Add comment about the required key in the CABundle secret

Config
------
Add configuration flag for the eorker setup container image name

Unit Tests
----------
Add tests to the new code

Sample
------
Rename the worker to avoid conflict with node
Add caBundleSecretReference

### CODE REVIEW

File: Dockerfile.WorkerSetupJob
Comment (bardielle) [inline Dockerfile.WorkerSetupJob:12]: I think it's better to use a specific version that's working for you in the lab environment

File: Dockerfile.WorkerSetupJob
Comment (bardielle) [inline Dockerfile.WorkerSetupJob:13]: in a disconnected env this should be replaced, doesn't it?

File: controllers/osbuildenvconfig_controller.go
Comment (bardielle) [inline controllers/osbuildenvconfig_controller.go:347]: The update function becomes to be a too long function that includes too many update cases
I think it's better to export each update to a separate function - first part for ensure composer resources and the second part will be for the workers

File: controllers/osbuildenvconfig_controller.go
Comment (bardielle) [inline controllers/osbuildenvconfig_controller.go:682]: more logs is needed here

File: Dockerfile.WorkerSetupJob
Comment (ygalblum) [reply Dockerfile.WorkerSetupJob:13]: No, we build this image ourselves and push it to our container registry.

File: controllers/osbuildenvconfig_controller.go
Comment (bardielle) [inline controllers/osbuildenvconfig_controller.go:348]: those 6 lines can be exposed to a function - for example if you'll decide to requeueAfter a period time,
something like that: 
func (r *OSBuildEnvConfigReconciler)  <WHATEVER-NAME>(created boolean, err error, msg string) (ctrl.Result, error){
       if err != nil {
		return ctrl.Result{Requeue: true}, err
	} else if created {
		reqLogger.Info(msg)
		return ctrl.Result{Requeue: true}, nil
	}
}

File: controllers/osbuildenvconfig_controller.go
Comment (bardielle) [inline controllers/osbuildenvconfig_controller.go:773]: Maybe better to create a type for that struct? 
not sure, maybe for testing it will be easier?

File: controllers/osbuildenvconfig_controller.go
Comment (bardielle) [inline controllers/osbuildenvconfig_controller.go:818]: again - I think it will be better to export is as a type

File: controllers/osbuildenvconfig_controller.go
Comment (bardielle) [inline controllers/osbuildenvconfig_controller.go:834]: again - I think it will be better to export is as a type

File: controllers/osbuildenvconfig_controller.go
Comment (bardielle) [inline controllers/osbuildenvconfig_controller.go:848]: it seems the same struct as above...

File: controllers/osbuildenvconfig_controller.go
Comment (bardielle) [inline controllers/osbuildenvconfig_controller.go:892]: again - I think it will be better to export is as a type

File: controllers/osbuildenvconfig_controller.go
Comment (ygalblum) [reply controllers/osbuildenvconfig_controller.go:848]: Not sure I understand your comment here. Anyhow I'll move the struct definition from inline to type, so the change might make this comment obsolete

File: Dockerfile.WorkerSetupJob
Comment (ygalblum) [reply Dockerfile.WorkerSetupJob]: Done

File: controllers/osbuildenvconfig_controller.go
Comment (ygalblum) [reply controllers/osbuildenvconfig_controller.go:776]: Why here? This is a recurring pattern in which the error, if happens, is printed at the top of the calling stack

File: controllers/osbuildenvconfig_controller.go
Comment (ygalblum) [reply controllers/osbuildenvconfig_controller.go]: Done

File: controllers/osbuildenvconfig_controller.go
Comment (ygalblum) [reply controllers/osbuildenvconfig_controller.go]: I don't really see the benefit, but I made the change

File: controllers/osbuildenvconfig_controller.go
Comment (ygalblum) [reply controllers/osbuildenvconfig_controller.go]: I don't really see the benefit, but I made the change

File: controllers/osbuildenvconfig_controller.go
Comment (ygalblum) [reply controllers/osbuildenvconfig_controller.go]: I don't really see the benefit, but I made the change

**Review by bardielle — COMMENTED** [review]
it looks good! great job
After fixing the build creation, ping me to approve that PR

**Comment by ygalblum** [comment]
@bardielle With the new commits, I am able to build an image using cUrl. So, I think we can merge this PR and if needed following the E2E tests, open another one

**Review by bardielle — APPROVED** [review]

**Pull request merged**
